integration_tests/test_search.py

```diff
@@ -1,25 +1,2078 @@
+# encoding: utf-8
+import logging
+import urllib
 from nose.tools import (
-    set_trace,
     eq_,
+    set_trace,
+)
+import os
+import re
+import json
+from core.model import (
+    production_session,
+    Library,
 )
-from urllib import urlopen
-import feedparser
+from core.lane import Pagination
+from core.external_search import (
+    ExternalSearchIndex,
+    Filter
+)
+
+
+# problems identified in ES6:
+# 'modern romance' - exact title match isn't promoted
+# (but it also doesn't work in ES1 on a real site)
+
+# summary is boosted over subtitle
+
+
+
+# # In Elasticsearch 6, the exact author match that doesn't
+# # mention 'biography' is boosted above a book that
+# # mentions all three words in its title.
+# order = [
+#     self.biography_of_peter_graves, # title + genre 'biography'
+#     self.book_by_peter_graves,      # author (no 'biography')
+#     self.behind_the_scenes,         # all words match in title
+#     self.book_by_someone_else,      # match across fields (no 'biography')
+# ]
+
+class Searcher(object):
+    def __init__(self, library, index):
+        self.library = library
+        self.filter = Filter(collections=self.library)
+        self.index = index
+
+    def query(self, query, pagination):
+        return self.index.query_works(
+            query, filter=self.filter, pagination=pagination,
+            debug=True, return_raw_results=True
+        )
+
+
+class Evaluator(object):
+    """A class that knows how to evaluate search results."""
+
+    log = logging.getLogger("Search evaluator")
+
+    def __init__(self, **kwargs):
+        self.kwargs = dict()
+        for k, v in kwargs.items():
+            if isinstance(v, basestring):
+                v = v.lower()
+            self.kwargs[k] = v
+        for k, v in self.kwargs.items():
+            setattr(self, k, v)
+
+    def evaluate(self, hits):
+        """Raise an AssertionError if the search results are so bad that the
+        test should fail.
+        """
+        self.evaluate_search(hits)
+        first = hits[0]
+        self.evaluate_first(first)
+        self.evaluate_hits(hits)
+
+    def evaluate_search(self, hits):
+        """Evaluate the search itself."""
+        # By default, a search passes the check if it returns anything
+        # as opposed to returning nothing.
```

Comment:
There may be cases where we want to return nothing, but maybe not with NYPL's collection since it's so large.

Reply:
I added a set of tests that validate that you get no results if you type a bunch of junk into the search box.